### ISSUE TITLE
Creating an empty user ("") breaks the user database

### ISSUE THREAD

**Issue by hillu** [google/grr#480]
Starting off with an installation of the Debian package (grr-server 3.1.0.2+dfsg-1), configured to use the SQLite3 datastore and the default admin user as set up by `grr_config_updater`, let's check the user database:
```
$ grr_config_updater show_user
Using configuration <GrrConfigManager  file="/usr/share/grr-server/install_data/etc/grr-server.yaml"  file="/etc/grr/server.local.yaml" >

Username: admin
Labels: admin
Password: set
$ 
```
I create a user name called <empty string>:
```
$ grr_config_updater add_user ""
Using configuration <GrrConfigManager  file="/usr/share/grr-server/install_data/etc/grr-server.yaml"  file="/etc/grr/server.local.yaml" >
Please enter password for user '': 
Added user .
```
Now the user database is broken:
```
$ grr_config_updater show_user
Using configuration <GrrConfigManager  file="/usr/share/grr-server/install_data/etc/grr-server.yaml"  file="/etc/grr/server.local.yaml" >
Traceback (most recent call last):
  File "/usr/bin/grr_config_updater", line 11, in <module>
    load_entry_point('grr-response-server==3.1.0.post2', 'console_scripts', 'grr_config_updater')()
  File "/usr/lib/python2.7/dist-packages/grr/lib/distro_entry.py", line 75, in ConfigUpdater
    flags.StartMain(config_updater.main)
  File "/usr/lib/python2.7/dist-packages/grr/lib/flags.py", line 121, in StartMain
    main([sys.argv[0]])
  File "/usr/lib/python2.7/dist-packages/grr/tools/config_updater.py", line 925, in main
    ShowUser(flags.FLAGS.username, token=token)
  File "/usr/lib/python2.7/dist-packages/grr/tools/config_updater.py", line 265, in ShowUser
    for user in fd.OpenChildren():
AttributeError: 'GRRUser' object has no attribute 'OpenChildren'
```
I seem to be able to remove that <empty string> user:
```
$ grr_config_updater delete_user ""
Using configuration <GrrConfigManager  file="/usr/share/grr-server/install_data/etc/grr-server.yaml"  file="/etc/grr/server.local.yaml" >
User  has been deleted.
```
But wait! Now, my admin user is also gone:
```
$ grr_config_updater show_user
Using configuration <GrrConfigManager  file="/usr/share/grr-server/install_data/etc/grr-server.yaml"  file="/etc/grr/server.local.yaml" >
```
Checking in the SQLite table shows that all entries (even those for `aff4:/users/admin`) have been removed:
```
$ sqlite3 datastore/users.sqlite 'select * from tbl;'
```
I realize that the SQLite datastore may not be a high priority thing for the core developers, so I'd appreciate some advice on where to fix this. (I got the feeling that there are at least two issues, namely a datastore issue and a missing check for illegal / nonsensical user names.)

**Comment by mbushkov** [comment]
Hi Hilko,
As far as I can tell, this behavior is not specific to SQLite datastore.
It's clearly a bug: grr_config_updater should reject invalid user names.

What happens is the following: user objects are created in aff4:/users/
AFF4 namespace. So normally user objects are identified by URNs like:
aff4:/users/a, aff4:/users/b, etc. When you create an user object with an
empty name, it effectively overwrites the "aff4:/users" path. aff4:/users/
stops being an AFF4Volume object (standard type for 'folders' in AFF4) and
becomes GRRUser object. As AFF4Volume and GRRUser objects have vastly
different characteristics and methods, so the system breaks.

When you try to delete a user with an empty name, it effectively deletes
aff4:/users/. The deletion happens recursively, thus all users are deleted.

As far as I understand, the code to be fixed is in maintenance_utils:
https://github.com/google/grr/blob/master/grr/lib/maintenance_utils.py#L328.
See AddUser/UpdateUser/DeleteUser functions.

Quick question, though: config_updater is meant to be used by GRR admins
only. Why creating a user with an empty name is a concern?

Cheers,
Misha

**Comment by hillu** [comment]
Hi Misha,

thank you for your direction, I'll see if I can get this fixed. 

It was my fear that the problem was not tied to the specific datastore.

> Quick question, though: config_updater is meant to be used by GRR admins
> only. Why creating a user with an empty name is a concern?

I am trying to integrate GRR with another software product and for that
I need to inspect and modify GRR users. This is also pretty much the
answer to your question you asked in #478. (As you may have guessed, I
hosed my user database when testing some of my integration code.)

Cheers,
-Hilko

**Comment by hillu** [comment]
Hi Misha,

> What happens is the following: user objects are created in aff4:/users/
> AFF4 namespace. So normally user objects are identified by URNs like:
> aff4:/users/a, aff4:/users/b, etc. When you create an user object with an
> empty name, it effectively overwrites the "aff4:/users" path.

So, this is where the breakage happens in AddUser:

```
fd = aff4.FACTORY.Create("aff4:/users/%s" % username,
                         users.GRRUser,
                         mode="rw",
                         token=token)
```

My first instinct was to look for a flag like O_EXCL for open(2) to make
aff4.Factory.Create() fail if a named object already exists.
Unfortunately there seems to be no such flag. :-(

Cheers,
-Hilko

**Comment by grrrrrrrrr** [comment]
Fixed by #481
